agent/rpc/subscribe/auth.go

```diff
@@ -0,0 +1,40 @@
+package subscribe
+
+import (
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/consul/stream"
+)
+
+// EnforceACL takes an acl.Authorizer and returns the decision for whether the
+// event is allowed to be sent to this client or not.
+func enforceACL(authz acl.Authorizer, e stream.Event) acl.EnforcementDecision {
+	switch {
+	case e.IsEndOfSnapshot(), e.IsEndOfEmptySnapshot():
+		return acl.Allow
+	}
+
+	switch p := e.Payload.(type) {
+	case state.EventPayloadCheckServiceNode:
+		csn := p.Value
+		if csn.Node == nil || csn.Service == nil || csn.Node.Node == "" || csn.Service.Service == "" {
+			return acl.Deny
+		}
+
+		// TODO: what about acl.Default?
```

Comment:
Calling out this TODO. Should this code expect to handle `acl.Default`? I guess currently it assumes default deny is ACLs are enabled, which I think I've heard is the expectation, but I thought I should double check.

Reply:
`acl.Default` is to allow chaining Authorizers and delegating the decision to something further on in the chain rather than making an explicit allow/deny decision. The `acl.ChainedAuthorizer` type is what Consul creates and it specially processes the `acl.Default` decision from a child authorizer and proceeds with the next authorizer in the chain.

https://github.com/hashicorp/consul/blob/bf2b52f88061520c4c5b996df91d40023537e46b/acl/chained_authorizer.go#L26-L34

Reply:
So basically the top level authorizer should return `Allow` or things should be denied.

Reply:
Also as far as the ACL package is concerned the true default decision is always `Deny`. From the overall product perspective the `ChainedAuthorizer` has to be created with the final authorizer in the chain as a `StaticAuthorizer` set to implement the products `default_policy = "allow"` mode.